fourier/rfft.go

```diff
@@ -0,0 +1,1151 @@
+//           r[i] = r[0] plus the sum from k=1 to k=(n-1)/2 of
+//             2*r(2*k-1)*cos(k*i*2*pi/n)
+//             -2*r(2*k)*sin(k*i*2*pi/n)
+//
+//  This transform is unnormalized since a call of rfftf
+//  followed by a call of rfftb will multiply the input
+//  sequence by n.
+//
+//  work   Contains results which must not be destroyed between
+//         calls of rfftf or rfftb.
+//  ifac   Contains results which must not be destroyed between
+//         calls of rfftf or rfftb.
+func rfftb(n int, r, work []float64, ifac []int) {
+	if len(r) < n {
+		panic("fourier: short sequence")
+	}
+	if len(work) < 2*n {
+		panic("fourier: short work")
+	}
+	if len(ifac) < 15 {
+		panic("fourier: short ifac")
+	}
+	if n == 1 {
+		return
+	}
+	rfftb1(n, r[:n], work[:n], work[n:2*n], ifac[:15])
+}
+
+func rfftb1(n int, c, ch, wa []float64, ifac []int) {
+	nf := ifac[1]
+	na := false
+	l1 := 1
+	iw := 0
+
+	for k1 := 1; k1 <= nf; k1++ {
+		ip := ifac[k1+1]
+		l2 := ip * l1
+		ido := n / l2
+		idl1 := ido * l1
+
+		switch ip {
+		case 4:
+			ix2 := iw + ido
+			ix3 := ix2 + ido
+			if na {
+				radb4(ido, l1, ch, c, wa[iw:], wa[ix2:], wa[ix3:])
+			} else {
+				radb4(ido, l1, c, ch, wa[iw:], wa[ix2:], wa[ix3:])
+			}
+			na = !na
+		case 2:
+			if na {
+				radb2(ido, l1, ch, c, wa[iw:])
+			} else {
+				radb2(ido, l1, c, ch, wa[iw:])
+			}
+			na = !na
+		case 3:
+			ix2 := iw + ido
+			if na {
+				radb3(ido, l1, ch, c, wa[iw:], wa[ix2:])
+			} else {
+				radb3(ido, l1, c, ch, wa[iw:], wa[ix2:])
+			}
+			na = !na
+		case 5:
+			ix2 := iw + ido
+			ix3 := ix2 + ido
+			ix4 := ix3 + ido
+			if na {
+				radb5(ido, l1, ch, c, wa[iw:], wa[ix2:], wa[ix3:], wa[ix4:])
+			} else {
+				radb5(ido, l1, c, ch, wa[iw:], wa[ix2:], wa[ix3:], wa[ix4:])
+			}
+			na = !na
+		default:
+			if na {
+				radbg(ido, ip, l1, idl1, ch, ch, ch, c, c, wa[iw:])
+			} else {
+				radbg(ido, ip, l1, idl1, c, c, c, ch, ch, wa[iw:])
+			}
+			if ido == 1 {
+				na = !na
+			}
+		}
+
+		l1 = l2
+		iw += (ip - 1) * ido
+	}
+
+	if na {
+		for i := 0; i < n; i++ {
+			c[i] = ch[i]
+		}
+	}
+}
+
+func radb2(ido, l1 int, cc, ch, wa1 []float64) {
+	cc3 := newThreeArray(ido, 2, l1, cc)
+	ch3 := newThreeArray(ido, l1, 2, ch)
+
+	for k := 0; k < l1; k++ {
+		ch3.set(0, k, 0, cc3.at(0, 0, k)+cc3.at(ido-1, 1, k))
+		ch3.set(0, k, 1, cc3.at(0, 0, k)-cc3.at(ido-1, 1, k))
+	}
+
+	if ido < 2 {
+		return
+	}
+	if ido > 2 {
+		idp2 := ido + 1
+		for k := 0; k < l1; k++ {
+			for i := 2; i < ido; i += 2 {
+				ic := idp2 - (i + 1)
+				ch3.set(i-1, k, 0, cc3.at(i-1, 0, k)+cc3.at(ic-1, 1, k))
+				tr2 := cc3.at(i-1, 0, k) - cc3.at(ic-1, 1, k)
+				ch3.set(i, k, 0, cc3.at(i, 0, k)-cc3.at(ic, 1, k))
+				ti2 := cc3.at(i, 0, k) + cc3.at(ic, 1, k)
+				ch3.set(i-1, k, 1, wa1[i-2]*tr2-wa1[i-1]*ti2)
+				ch3.set(i, k, 1, wa1[i-2]*ti2+wa1[i-1]*tr2)
+			}
+		}
+
+		if ido%2 == 1 {
+			return
+		}
+	}
+	for k := 0; k < l1; k++ {
+		ch3.set(ido-1, k, 0, cc3.at(ido-1, 0, k)+cc3.at(ido-1, 0, k))
```

Comment:
Shouldn't these just be `2*cc3.at(` looks like the same call twice.

Reply:
Yes, damn fortraners.